app/src/main/java/com/nutomic/syncthingandroid/util/ConfigXml.java

```diff
@@ -110,9 +138,69 @@ public String getUserName() {
      */
     @SuppressWarnings("SdCardPath")
     public void updateIfNeeded() {
-        Log.i(TAG, "Checking for needed config updates");
         boolean changed = false;
+
+        /* Get preference - PREF_USE_FOLDER_OBSERVER */
+        boolean prefUseFolderObserver = false;
+        mPreferences = PreferenceManager.getDefaultSharedPreferences(mContext);
+        prefUseFolderObserver = mPreferences.getBoolean(Constants.PREF_USE_FOLDER_OBSERVER, false);
+
+        /* Read existing config version */
+        int iConfigVersion = Integer.parseInt(mConfig.getDocumentElement().getAttribute("version"));
+        Log.i(TAG, "Found existing config version " + Integer.toString(iConfigVersion));
+
+        /* Get refs to important config objects */
         NodeList folders = mConfig.getDocumentElement().getElementsByTagName("folder");
+
+        /* Check if we have to do manual migration from version X to Y */
+        if (iConfigVersion == 27) {
+          /* fsWatcher transition - https://github.com/syncthing/syncthing/issues/4882 */
+          Log.i(TAG, "Migrating config version " + Integer.toString(iConfigVersion) + " to 28 ...");
+
+          /* Enable fsWatcher for all folders */
+          for (int i = 0; i < folders.getLength(); i++) {
+            Element r = (Element) folders.item(i);
+
+            // Enable "fsWatcherEnabled" attribute.
+            if (!r.hasAttribute("fsWatcherEnabled") ||
+                    !Boolean.parseBoolean(r.getAttribute("fsWatcherEnabled"))) {
+              Log.i(TAG, "Set 'fsWatcherEnabled' on folder " + r.getAttribute("id"));
+              r.setAttribute("fsWatcherEnabled", Boolean.toString(true));
+              changed = true;
+            }
+          }
```

Comment:
This section that tweaks the fsWatcherEnabled can go away, and just be handled below in the if (prefUseFolderObserver) branch, removing the if and always doing
```
r.setAttribute("fsWatcherEnabled", Boolean.toString(!prefUseFolderObserver));
```

Reply:
Much more easier! I was little blinded by Nutomics strategy only writing the config back if it was changed and closely tracking changes to the config - but you're right - assuming we do a migration from version 27>28 once, we are sure we change the config during this step and we must write it back. So indeed the extra code for "only logging and saving if the config changed" can be removed here.

Reply:
Setting it the opposite of "prefUseFolderObserver" has a disadvantage from my point of view as the users of the web UI are forced to live with or without the file watcher. On my point of view, there should be more liberty by:
a) Forcing file watcher on during one-time migration for all folders and then leaving the decision of the filewatcher on/off up to the user
b) If FolderObserver is enabled in the app settings UI, the user is "helped" in preventing a manual misconfiguration by force-turning off the fsWatcher for all folders. This is also stated in the explanation text under the experimental option checkbox.